parity/main.rs

```diff
@@ -160,61 +168,120 @@ fn start() -> Result<String, String> {
 	}
 
 	let cmd = try!(conf.into_command());
-	execute(cmd)
+	execute(cmd, can_restart)
 }
 
+#[cfg(not(feature="stratum"))]
+fn stratum_main(_: &mut HashMap<String, fn()>) {}
+
 #[cfg(feature="stratum")]
-mod stratum_optional {
-	pub fn probably_run() -> bool {
-		// just redirect to the stratum::main()
-		if ::std::env::args().nth(1).map_or(false, |arg| arg == "stratum") {
-			super::stratum::main();
-			true
-		}
-		else { false }
-	}
+fn stratum_main(alt_mains: &mut HashMap<String, fn()>) {
+	alt_mains.insert("stratum".to_owned(), stratum::main);
 }
 
-#[cfg(not(feature="stratum"))]
-mod stratum_optional {
-	pub fn probably_run() -> bool {
-		false
-	}
+#[cfg(not(feature="ipc"))]
+fn sync_main(_: &mut HashMap<String, fn()>) {}
+
+#[cfg(feature="ipc")]
+fn sync_main(alt_mains: &mut HashMap<String, fn()>) {
+	alt_mains.insert("sync".to_owned(), sync::main);
 }
 
-#[cfg(not(feature="ipc"))]
-fn sync_main() -> bool {
-	false
+// TODO: merge with version in Updater.
+fn updates_path(name: &str) -> PathBuf {
+	let mut dest = PathBuf::from(env::home_dir().unwrap().to_str().expect("env filesystem paths really should be valid; qed"));
+	dest.push(".parity-updates");
+	dest.push(name);
+	dest
 }
 
-#[cfg(feature="ipc")]
-fn sync_main() -> bool {
-	// just redirect to the sync::main()
-	if std::env::args().nth(1).map_or(false, |arg| arg == "sync") {
-		sync::main();
-		true
+fn latest_exe_path() -> Option<PathBuf> {
+	File::open(updates_path("latest")).ok()
+		.and_then(|mut f| { let mut exe = String::new(); f.read_to_string(&mut exe).ok().map(|_| updates_path(&exe)) })
+} 
+
+// Starts ~/.parity-updates/parity and returns the code it exits with.
+fn run_parity() -> Option<i32> {
+	use ::std::ffi::OsString;
+	let prefix = vec![OsString::from("--can-restart"), OsString::from("--force-direct")];
+	latest_exe_path().and_then(|exe| process::Command::new(exe)
```

Comment:
AKA for auto-updating to be done securely it should be done with super-user privs and written to a write-protected updates directory, dropping privileges immediately after that.